src/syntaxtree.c

```diff
@@ -0,0 +1,1459 @@
+    return result;
+}
+
+static inline Obj SyntaxTreeExpr(Expr expr)
+{
+    return AllExpressions[TNUM_EXPR(expr)].compiler(expr);
+}
+
+Obj SyntaxTreeUnknownExpr(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("UnknownExpr");;
+    AssPRec(result, RNamName("tnum"), INTOBJ_INT(TNUM_EXPR(expr)));
+    return result;
+}
+
+Obj SyntaxTreeUnknownBool(Expr expr)
+{
+    /* compile the expression and check that the value is boolean          */
+    /* TODO: Check boolean? */
+    return SyntaxTreeExpr(expr);
+}
+
+extern Obj SyntaxTreeRefGVarFopy (Expr expr);
+
+Obj SyntaxTreeFunccall(Expr expr)
+{
+    Obj result;
+    Obj func;
+    Obj args, argi;
+    UInt narg, i;
+
+    result = NewSyntaxTreeNode("Funccall");;
+    /* TODO: If this is a gvar ref, put name? */
+    func = SyntaxTreeExpr( FUNC_CALL(expr) );
+
+    /*
+    if ( TNUM_EXPR( FUNC_CALL(expr) ) == T_REF_GVAR ) {
+    } */
+    AssPRec(result, RNamName("function"), func);
+
+    /* compile the argument expressions                                    */
+    narg = NARG_SIZE_CALL(SIZE_EXPR(expr));
+    args = NEW_PLIST(T_PLIST, narg);
+    SET_LEN_PLIST(args, narg);
+
+    for ( i = 1; i <= narg; i++ ) {
+        argi = SyntaxTreeExpr( ARGI_CALL( expr, i ) );
+        SET_ELM_PLIST(args, i, argi);
+        CHANGED_BAG(args);
+    }
+    AssPRec(result, RNamName("args"), args);
+    return result;
+}
+
+/* TODO: FuncCall options */
+Obj SyntaxTreeFunccallOpts(Expr expr)
+{
+    Obj result;
+    Obj opts;
+
+    result = NewSyntaxTreeNode("FunccallOpts");;
+    opts = SyntaxTreeExpr(ADDR_STAT(expr)[0]);
+    AssPRec(result, RNamName("opts"), opts);
+
+    return result;
+}
+
+static Obj SyntaxTreeFunc(Obj func);
+
+/* TODO: again? */
+Obj SyntaxTreeFuncExpr(Expr expr)
+{
+    Obj result;
+    Obj fexs;
+    Obj fexp;
+    Obj func;
+
+    result = NewSyntaxTreeNode("FuncExpr");;
+
+    fexs = FEXS_FUNC(CURR_FUNC);
+    fexp = ELM_PLIST(fexs, ((Int *)ADDR_EXPR(expr))[0]);
+    func = SyntaxTreeFunc(fexp);
+
+    AssPRec(result, RNamName("func"), func);
+
+    return result;
+}
+
+Obj SyntaxTreeIntExpr(Expr expr)
+{
+    Obj result;
+    Obj value;
+    Int siz;
+    UInt typ;
+    Int i;
+
+    result = NewSyntaxTreeNode("IntExpr");;
+
+    if(IS_INTEXPR(expr)) {
+        value = OBJ_INTEXPR(expr);
+    } else {
+        siz = SIZE_EXPR(expr) - sizeof(UInt);
+        typ = *(UInt *)ADDR_EXPR(expr);
+        value = C_MAKE_INTEGER_BAG(siz, typ);
+        for ( i = 0; i < siz/INTEGER_UNIT_SIZE; i++ ) {
+#if INTEGER_UNIT_SIZE == 2
+            C_SET_LIMB2(value, i, ((UInt2 *)((UInt *)ADDR_EXPR(expr) + 1))[i]);
+#elif INTEGER_UNIT_SIZE == 4
+            C_SET_LIMB4(value, i, ((UInt4 *)((UInt *)ADDR_EXPR(expr) + 1))[i]);
+#elif INTEGER_UNIT_SIZE == 8
+            C_SET_LIMB8(value, i, ((UInt8 *)((UInt *)ADDR_EXPR(expr) + 1))[i]);
+#else
+            #error unsupported INTEGER_UNIT_SIZE
+#endif
+        }
+        if (siz <= 8) {
+            value = C_NORMALIZE_64BIT(value);
```

Comment:
Oops, I wasn't even aware of `C_NORMALIZE_64BIT` and the above code (which comes from the compiler, I think?). I'd prefer if all integer normalization was handled by code in `gmpints.c`, and if code outside assumed as little as possible about the internal format of the integer code, so that future changes require less code changes outside of `gmpints.c`. I'll try to understand this code and see if I can suggest an alternative...